src/Libraries/Microsoft.Extensions.AI.Abstractions/ChatCompletion/StreamingChatCompletionUpdateExtensions.cs

```diff
@@ -0,0 +1,212 @@
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+#if NET
+using System.Runtime.InteropServices;
+#endif
+using System.Text;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Shared.Diagnostics;
+
+#pragma warning disable S109 // Magic numbers should not be used
+#pragma warning disable S127 // "for" loop stop conditions should be invariant
+
+namespace Microsoft.Extensions.AI;
+
+/// <summary>
+/// Provides extension methods for working with <see cref="StreamingChatCompletionUpdate"/> instances.
+/// </summary>
+public static class StreamingChatCompletionUpdateExtensions
+{
+    /// <summary>Combines <see cref="StreamingChatCompletionUpdate"/> instances into a single <see cref="ChatCompletion"/>.</summary>
+    /// <param name="updates">The updates to be combined.</param>
+    /// <param name="coalesceContent">
+    /// <see langword="true"/> to attempt to coalesce contiguous <see cref="AIContent"/> items, where applicable,
+    /// into a single <see cref="AIContent"/>, in order to reduce the number of individual content items that are included in
+    /// the manufactured <see cref="ChatMessage"/> instances. When <see langword="false"/>, the original content items are used.
+    /// The default is <see langword="true"/>.
+    /// </param>
+    /// <returns>The combined <see cref="ChatCompletion"/>.</returns>
+    public static ChatCompletion ToChatCompletion(
+        this IEnumerable<StreamingChatCompletionUpdate> updates, bool coalesceContent = true)
+    {
+        _ = Throw.IfNull(updates);
+
+        ChatCompletion completion = new([]);
+        Dictionary<int, ChatMessage> messages = [];
+
+        foreach (var update in updates)
+        {
+            ProcessUpdate(update, messages, completion);
+        }
+
+        AddMessagesToCompletion(messages, completion, coalesceContent);
+
+        return completion;
+    }
+
+    /// <summary>Combines <see cref="StreamingChatCompletionUpdate"/> instances into a single <see cref="ChatCompletion"/>.</summary>
+    /// <param name="updates">The updates to be combined.</param>
+    /// <param name="coalesceContent">
+    /// <see langword="true"/> to attempt to coalesce contiguous <see cref="AIContent"/> items, where applicable,
+    /// into a single <see cref="AIContent"/>, in order to reduce the number of individual content items that are included in
+    /// the manufactured <see cref="ChatMessage"/> instances. When <see langword="false"/>, the original content items are used.
+    /// The default is <see langword="true"/>.
+    /// </param>
+    /// <param name="cancellationToken">The <see cref="CancellationToken"/> to monitor for cancellation requests. The default is <see cref="CancellationToken.None"/>.</param>
+    /// <returns>The combined <see cref="ChatCompletion"/>.</returns>
+    public static Task<ChatCompletion> ToChatCompletionAsync(
```

Comment:
This approach seems cleaner, but at the same time doesn't it negate the purpose of streaming completions? You wouldn't be able to read anything until the enumeration has finished. A mechanism for injecting side-effects as results arrive (e.g. by way of a `Select` or `Tee` method) becomes almost mandatory.

Reply:
But since we don't have a readily available Linq for IAE, perhaps it makes sense to bake functionality into this method, by way of an optional `Action<StreamingChatCompletionUpdate>? onCompletionUpdate` callback parameter.

Reply:
The enumerable one isn't "blocking". The intent is it works on an already created collection, eg
```C#
List<StreamingChatCompletionUpdate> updates = [];
await foreach (var update in client.CompleteStreamingAsync(messages))
{
    Console.Write(update);
    updates.Add(update);
}
messages.Add(updates.ToChatCompletion().Message);
```

And the async equivalent is primarily so that a client can implement its non-streaming implementation in terms of its streaming, eg
```C#
public Task<ChatCompletion> CompleteAsync(
        IList<ChatMessage> chatMessages, ChatOptions? options = null, CancellationToken cancellationToken = default) =>
    CompleteStreamingAsync(chatMessages, options, cancellationToken).ToChatCompletionAsync();
```